### PR TITLE
feat: download artifacts

### PR DESCRIPTION
**Pull request by ajaypratap003**
Before delete service registry instance user can download artifacts from delete modal by click on link.
Jira: https://issues.redhat.com/browse/MGDSTRM-4979

image

### CODE REVIEW

**Comment by ajaypratap003** [comment]
@EricWittmann  @pmuir Please review this PR.

**Review by carlesarnal — CHANGES_REQUESTED** [review]
Please address or reply to the comment about the location of the `downloadArtifacts` function.

File: ui/src/services/groups/groups.service.ts [hunk 1]
```diff
@@ -241,6 +241,18 @@ export class GroupsService extends BaseService {
         return this.httpDelete(endpoint);
     }
 
+    public downloadArtifacts(): Promise<string> {
```
[the remark this answers is not in the document]
Comment (ajaypratap003) [reply ui/src/services/groups/groups.service.ts:244]: @carlesarnal I'm okay to move in `admin.service.ts` if you want.

File: ui/src/app/pages/DownloadArtifacts/DownloadArtifactsFederated.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+import { DownloadArtifacts, DownloadArtifactsProps } from "./DownloadArtifacts";
+import { FederatedPageProps, FederatedUtils } from "../../federated";
+
+export interface DownloadArtifactsFederatedProps
+  extends DownloadArtifactsProps,
+    FederatedPageProps {}
+
+export default class FederatedArtifactsPage extends DownloadArtifacts {
```
Comment (EricWittmann) [inline ui/src/app/pages/DownloadArtifacts/DownloadArtifactsFederated.tsx:25]: Copy/paste error?

**Comment by EricWittmann** [comment]
This introduces a new data plane page, but it seems to just be a page with a single button on it.  Can you explain how you see this page working and perhaps include a screenshot in the PR description (if you have one that is relevant)?

I'm trying to figure out how I use this upstream.

**Comment by ajaypratap003** [comment]
> This introduces a new data plane page, but it seems to just be a page with a single button on it. Can you explain how you see this page working and perhaps include a screenshot in the PR description (if you have one that is relevant)?
> 
> I'm trying to figure out how I use this upstream.

image

**Comment by EricWittmann** [comment]
Thank you for the screenshot.  That is a modal dialog in the control plane which provides a link to downloading the ZIP file with all the artifacts.  What specifically happens when the user clicks that link?  Are they taken to a page with the download button on it?

I would expect that clicking the link just downloads the .zip directly.  I realize that's actually potentially tricky, since it means making an authenticated call to a REST API and then saving the response payload as a file to local disk.

@pmuir any thoughts on this?

File: ui/src/app/pages/DownloadArtifacts/DownloadArtifactsFederated.tsx (hunk 2 above)
Comment (EricWittmann) [reply ui/src/app/pages/DownloadArtifacts/DownloadArtifactsFederated.tsx]: Yes.  Seems like it should be FederatedDownloadArtifacts or some such.

**Comment by EricWittmann** [comment]
Reviewing the code more thoroughly, I now better understand what this does.  It dynamically creates a link and then clicks it.  It does the REST API call and then uses this to make the download happen:

```
    const downloadUrl = window.URL.createObjectURL(
      new Blob([response], { type: "application/zip" })
    );
```

Is this a safe thing to do for potentially large ZIP files?

It's clever.  :)

File: ui/src/app/pages/DownloadArtifacts/DownloadArtifacts.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    return {};
+  }
+
+  private createDownloadLink = (response: string): void => {
+    const { instanceName = "artifacts" } = this.props;
+    const downloadUrl = window.URL.createObjectURL(
+      new Blob([response], { type: "application/zip" })
+    );
+
+    const link = document.createElement("a");
```
Comment (pmuir) [inline ui/src/app/pages/DownloadArtifacts/DownloadArtifacts.tsx:47]: Why are we manipulating the DOM?

**Comment by pmuir** [comment]
Shouldn't this be something the backend supports? Normally this is done by generating a temporary key that can be sent as a request parameter.

**Comment by EricWittmann** [comment]
Good idea - that's a better approach.  You're suggesting an authenticated call to `/admin/export` would result in a JSON response with a single-use URL that doesn't require authentication?

That will be interesting to implement in the multi-tenant registry, because there is some tenant authorization code that would fail if no credentials exist (we check the rh-org-id claim in the JWT to make sure the user is really allowed to access that tenant).

Would need to think about this a bit...

**Comment by pmuir** [comment]
Exactly - you make an authenticated call to something like `/admin/export` and it returns you a url like `/admin/export/download/foo/bar.zip?key=<some key>`. That key is then valid for (say) 1 hour, and you can use that url as many times as you like or from any client. You basically have a table of valid keys, that also has the artifact the key is valid for, and the time it was created. You can have some sort of cron to clean up the table regularly. You could embed the org id somehow eg. `/admin/export/download/<rh-org-id>/foo/bar.zip?key=<some key>`.

**Comment by EricWittmann** [comment]
Yes I agree this makes sense for browser scenarios.  +1 from me - I'll add this to the backlog.  

I'm not sure I want to accept this PR based on the approach to downloading the file.  Unless I'm mistaken, this will try to load the entire response into memory (as a blob) prior to prompting the user to save the file to disk.  If true, that's too risky given the unknown size of the ZIP.  And it has the potential of being a poor UX for the user - if there is no progress meter for a larger download.

Instead let's work on getting this done properly on the backend.

Note for the future: having the rh-org-id in the URL isn't really a solution to that problem.  The issue is that we specifically check the JWT to make sure the rh-org-id in the claim matches the org associated with the tenant.  It's an added security issue.  Sticking the ID in the URL means we could check it from there, but of course it will match without providing any real security.  What we really need to do is skip the org-id check if **no credentials** are provided.  That way normal authorization can be applied to individual endpoints.  Most endpoints would then fail with "user not authenticated", but some endpoints (those that don't require the user to be authenticated) can work.  This is something we should do anyway to support some other endpoints that don't require authentication.

**Comment by EricWittmann** [comment]
@Apicurio/developers any additional thoughts/objections?

**Comment by pmuir** [comment]
Ah sorry, I misunderstood what you are doing the rh-org-id validation for :-)

I think you can just skip this if they have the magic token, as they can on;y have got it from using their valid JWT.

**Comment by EricWittmann** [comment]
@ajaypratap003 I'm going to close this in favor of implementing an improved Export flow better designed for browser use-cases.

**Pull request closed without merging**